File: tests/algorithms/x/Pascal/maths/average_median.pas
{$mode objfpc}{$modeswitch nestedprocvars}
program Main;
uses SysUtils;
type IntArray = array of integer;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
procedure error(msg: string);
begin
  panic(msg);
end;
function _to_float(x: integer): real;
begin
  _to_float := x;
end;
function to_float(x: integer): real;
begin
  to_float := _to_float(x);
end;
procedure json(xs: array of real);
var i: integer;
begin
  write('[');
  for i := 0 to High(xs) do begin
    write(xs[i]);
    if i < High(xs) then write(', ');
  end;
  writeln(']');
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  nums: IntArray;
function bubble_sort(nums: IntArray): IntArray; forward;
function median(nums: IntArray): real; forward;
function bubble_sort(nums: IntArray): IntArray;
var
  bubble_sort_arr: array of integer;
  bubble_sort_n: integer;
  bubble_sort_i: integer;
  bubble_sort_j: integer;
  bubble_sort_a: integer;
  bubble_sort_b: integer;
begin
  bubble_sort_arr := nums;
  bubble_sort_n := Length(bubble_sort_arr);
  bubble_sort_i := 0;
  while bubble_sort_i < bubble_sort_n do begin
  bubble_sort_j := 0;
  while bubble_sort_j < (bubble_sort_n - 1) do begin
  bubble_sort_a := bubble_sort_arr[bubble_sort_j];
  bubble_sort_b := bubble_sort_arr[bubble_sort_j + 1];
  if bubble_sort_a > bubble_sort_b then begin
  bubble_sort_arr[bubble_sort_j] := bubble_sort_b;
  bubble_sort_arr[bubble_sort_j + 1] := bubble_sort_a;
end;
  bubble_sort_j := bubble_sort_j + 1;
end;
  bubble_sort_i := bubble_sort_i + 1;
end;
  exit(bubble_sort_arr);
end;
function median(nums: IntArray): real;
var
  median_sorted_list: IntArray;
  median_length_: integer;
  median_mid_index: integer;
begin
  median_sorted_list := bubble_sort(nums);
  median_length_ := Length(median_sorted_list);
  median_mid_index := median_length_ div 2;
  if (median_length_ mod 2) = 0 then begin
  exit(Double(median_sorted_list[median_mid_index] + median_sorted_list[median_mid_index - 1]) / 2);
end else begin
  exit(Double(median_sorted_list[median_mid_index]));
end;
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  writeln(FloatToStr(median([0])));
  writeln(FloatToStr(median([4, 1, 3, 2])));
  writeln(FloatToStr(median([2, 70, 6, 50, 20, 8, 4])));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.

File: tests/algorithms/x/Clojure/maths/average_median.clj
(ns main (:refer-clojure :exclude [bubble_sort median]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(defn toi [s]
  (Integer/parseInt (str s)))

(defn _fetch [url]
  {:data [{:from "" :intensity {:actual 0 :forecast 0 :index ""} :to ""}]})

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare bubble_sort median)

(def ^:dynamic bubble_sort_a nil)

(def ^:dynamic bubble_sort_arr nil)

(def ^:dynamic bubble_sort_b nil)

(def ^:dynamic bubble_sort_i nil)

(def ^:dynamic bubble_sort_j nil)

(def ^:dynamic bubble_sort_n nil)

(def ^:dynamic median_length nil)

(def ^:dynamic median_mid_index nil)

(def ^:dynamic median_sorted_list nil)

(defn bubble_sort [bubble_sort_nums]
  (binding [bubble_sort_a nil bubble_sort_arr nil bubble_sort_b nil bubble_sort_i nil bubble_sort_j nil bubble_sort_n nil] (try (do (set! bubble_sort_arr bubble_sort_nums) (set! bubble_sort_n (count bubble_sort_arr)) (set! bubble_sort_i 0) (while (< bubble_sort_i bubble_sort_n) (do (set! bubble_sort_j 0) (while (< bubble_sort_j (- bubble_sort_n 1)) (do (set! bubble_sort_a (nth bubble_sort_arr bubble_sort_j)) (set! bubble_sort_b (nth bubble_sort_arr (+ bubble_sort_j 1))) (when (> bubble_sort_a bubble_sort_b) (do (set! bubble_sort_arr (assoc bubble_sort_arr bubble_sort_j bubble_sort_b)) (set! bubble_sort_arr (assoc bubble_sort_arr (+ bubble_sort_j 1) bubble_sort_a)))) (set! bubble_sort_j (+ bubble_sort_j 1)))) (set! bubble_sort_i (+ bubble_sort_i 1)))) (throw (ex-info "return" {:v bubble_sort_arr}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn median [median_nums]
  (binding [median_length nil median_mid_index nil median_sorted_list nil] (try (do (set! median_sorted_list (bubble_sort median_nums)) (set! median_length (count median_sorted_list)) (set! median_mid_index (/ median_length 2)) (if (= (mod median_length 2) 0) (throw (ex-info "return" {:v (/ (double (+ (nth median_sorted_list median_mid_index) (nth median_sorted_list (- median_mid_index 1)))) 2.0)})) (throw (ex-info "return" {:v (double (nth median_sorted_list median_mid_index))})))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (println (str (median [0])))
      (println (str (median [4 1 3 2])))
      (println (str (median [2 70 6 50 20 8 4])))
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)
